jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/CompoundPool.java

```diff
@@ -0,0 +1,147 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+/**
+ * This implementation does not support multiplexing.
+ *
+ * @param <P> the type of the pooled objects
+ */
+public class CompoundPool<P> implements Pool<P>
+{
+    private final ConcurrentPool<P> concurrentPool;
+    private final Pool<P> secondBestPool;
+
+    public CompoundPool(ConcurrentPool.StrategyType strategyType, int maxSize, boolean cache)
+    {
+        if (maxSize <= ConcurrentPool.MAX_RECOMMENDED_SIZE)
+        {
+            this.concurrentPool = new ConcurrentPool<>(strategyType, maxSize, cache);
+            this.secondBestPool = new NullPool<>();
+        }
+        else
+        {
+            this.concurrentPool = new ConcurrentPool<>(strategyType, ConcurrentPool.MAX_RECOMMENDED_SIZE, cache);
+            this.secondBestPool = new QueuedPool<>(maxSize - ConcurrentPool.MAX_RECOMMENDED_SIZE);
+        }
+    }
+
+    @Override
+    public Entry<P> reserve()
+    {
+        Entry<P> reserved = concurrentPool.reserve();
+        if (reserved != null)
+            return reserved;
+        return secondBestPool.reserve();
```

Comment:
Stylewise, and naming-wise, I have a liking for:
```suggestion
        Entry<P> primary = primaryPool.reserve();
        return primary != null ? primary : secondaryPool.reserver();
```

Reply:
Done.